rfcs/2021-09-14-3910-schema-support.md

```diff
@@ -0,0 +1,472 @@
+always be an integer.
+
+It goes without saying that allowing this to exists means that the VRL compiler
+is entirely dependent on Vector providing the **correct** schema details for any
+given event. If it does not, and the received event is shaped differently from
+what the VRL compiler expected based on the schema definition it received,
+undefined runtime errors will occur when running the VRL program against the
+event.
+
+##### Output Schema Validation
+
+Additionally, if the operator specified the required output schema of the
+transform, Vector uses VRL's type checking to ensure that whatever the VRL
+program produces once it runs to completion, matches the expected output schema.
+
+If the output does not match the expected schema, Vector won't boot. This
+ensures that operators can update VRL programs while knowing that Vector will
+continue to validate the final output, preventing them from introducing a bug in
+the program that invalidates the schema adherence.
+
+```toml
+[transforms.remap]
+  type = "remap"
+  inputs = []
+
+  # Ensure any event produced by this transform adheres to my "custom" schema.
+  schema.out.type = "custom"
+  schema.out.file = "./my_schema.json"
+
+  # Given VRL's compile-time checking, we can check at Vector boot-time whether
+  # the output of this transform matches the expected schema. If it doesn't,
+  # Vector fails to boot.
+  source = """
+    . = { "foo": "bar" }
+  """
+```
+
+##### Output Schema Generation
+
+Even if the operator _does not_ specify a required output schema, Vector still
+built up the schema internally based on the output types of the VRL program.
+This schema can then be used by other transforms and sinks within the same
+pipeline.
+
+If the output of a VRL program matches a _known_ schema (e.g. `elasticsearch`),
+then events emitted by this transform automatically get that schema type
+attached to them, and if it doesn't match any known schemas, then they get
+a `custom` schema attached to them.
+
+Either way, using the `remap` transform in any pipeline automatically results in
+a schema being attached to the resulting events, courtesy of the VRL compiler.
+
+```toml
+[transforms.remap]
+  type = "remap"
+  inputs = []
+
+  # The output schema of this transform is defined to have a "foo" field with
+  # a string-value.
+  source = """
+    . = { "foo": "bar" }
+  """
+```
+
+#### Schema Conditions
+
+Finally, schemas can be used in "condition-based" transforms.
+
+A condition-based transform is one that performs an action on the received event
+based on a condition. Currently those include the `route`, `reduce`, `sample`,
+and `filter` transforms.
+
+Currently, these transforms take either the `vrl` condition type, or the
+deprecated `check_fields` type. After implementing this RFC, a new `schema` type
+is added, which allows applying conditional logic based on the schema attached
+to each individual event.
```

Comment:
Could a schema here be as generic as `json`?

Reply:
No, that's not a schema and should be validate upstream, ideally with the `codec` option.